lib/AST/Decl.cpp

```diff
@@ -3300,6 +3310,20 @@ ParamDecl::ParamDecl(ParamDecl *PD)
 
 
 /// \brief Retrieve the type of 'self' for the given context.
+Type DeclContext::getSelfTypeInContext() const {
+  // For a protocol or extension thereof, the type is 'Self'.
+  if (getAsProtocolOrProtocolExtensionContext()) {
+    // In the parser, generic parameters won't be wired up yet, just give up on
+    // producing a type.
+    if (getGenericParamsOfContext() == nullptr)
+      return Type();
+    return getProtocolSelf()->getArchetype();
+  }
+  return getDeclaredTypeInContext();
+}
+
+/// \brief Retrieve the type of 'self' for the given context.
+/// FIXME: Can this be integrated with getSelfTypeInContext above?
 static Type getSelfTypeForContext(DeclContext *dc) {
```

Comment:
The `static` implementation returns the *unbound* generic type. It has only one use, and I didn't want to disturb it.